// msw/lib/esm/styleStatusCode-deps.js
import { g as lib_5 } from './fetch-deps.js';

/**
 * Returns a parsed object from a given valid JSON string,
 * otherwise returns as text without parsing.
 */
function getJsonBody(body) {
    try {
        return JSON.parse(body);
    }
    catch (error) {
        return body;
    }
}

/**
 * Formats a mocked request for introspection in browser's console.
 */
function prepareRequest(req) {
    return Object.assign(Object.assign({}, req), { headers: req.headers.getAllHeaders() });
}

/**
 * Formats a mocked response for introspection in browser's console.
 */
function prepareResponse(res) {
    var _a;
    const resHeaders = lib_5(res.headers);
    return Object.assign(Object.assign({}, res), { 
        // Parse a response JSON body for preview in the logs
        body: ((_a = resHeaders.get('content-type')) === null || _a === void 0 ? void 0 : _a.includes('json')) ? getJsonBody(res.body)
            : res.body });
}

function getTimestamp() {
    const now = new Date();
    return [now.getHours(), now.getMinutes(), now.getSeconds()]
        .map(String)
        .map((chunk) => chunk.slice(0, 2))
        .map((chunk) => chunk.padStart(2, '0'))
        .join(':');
}

function styleStatusCode(status) {
    if (status < 300) {
        return 'color:#69AB32;';
    }
    if (status < 400) {
        return 'color:#F0BB4B;';
    }
    return 'color:#E95F5D;';
}

export { prepareResponse as a, getTimestamp as b, getJsonBody as g, prepareRequest as p, styleStatusCode as s };
