// @redwoodjs/dev-server/dist/main.js
#!/usr/bin/env node
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.getArgsForSide = void 0;

var _yargs = _interopRequireDefault(require("yargs"));

var _internal = require("@redwoodjs/internal");

var _http = require("./http");

var _watchApiSide = require("./watchApiSide");

var _awsLambda = require("./requestHandlers/awsLambda");

var _error = require("./error");

// TODO: Expand the sides once that concept is introduced.
const getArgsForSide = side => {
  const config = (0, _internal.getConfig)();
  const {
    port,
    host
  } = config[side];
  const paths = (0, _internal.getPaths)();
  return {
    port,
    host,
    paths: paths[side]
  };
};

exports.getArgsForSide = getArgsForSide;

const {
  side
} = _yargs.default.option('side', {
  default: 'api'
}).argv;

try {
  const {
    host,
    port,
    paths
  } = getArgsForSide(side);
  (0, _http.server)({
    requestHandler: _awsLambda.requestHandler
  }).listen(port, () => {
    console.log(`Running at 'http://${host}:${port}'`);
    console.log(`Watching files in '${paths.functions}'`);
    let startBuild = new Date().getTime();
    (0, _watchApiSide.watchFunctions)({
      paths,
      onChange: () => {
        startBuild = new Date().getTime();
        process.stdout.write('Change detected, building... ');
      },
      onImport: functions => {
        console.log(`Done. Took ${new Date().getTime() - startBuild}ms.`);
        (0, _http.setLambdaFunctions)(functions);
      },
      onException: async e => {
        console.log(await (0, _error.handleError)(e));
      }
    });
  });
} catch (e) {
  (0, _error.handleError)(e).then(m => {
    console.log(m);
    process.exit(1);
  });
}
