// @redwoodjs/api/dist/makeMergedSchema/rootSchema.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.resolvers = exports.schema = void 0;

var _graphqlTag = _interopRequireDefault(require("graphql-tag"));

var _graphqlIsoDate = require("graphql-iso-date");

var _graphqlTypeJson = _interopRequireWildcard(require("graphql-type-json"));

var _package = _interopRequireDefault(require("../../package.json"));

// @ts-ignore

/**
 * This adds scalar types for dealing with Date, Time, DateTime, and JSON.
 * This also adds a root Query type which is needed to start the GraphQL server on a fresh install.
 */
const schema = (0, _graphqlTag.default)`
  scalar Date
  scalar Time
  scalar DateTime
  scalar JSON
  scalar JSONObject

  type Redwood {
    version: String
    currentUser: JSON
  }

  type Query {
    redwood: Redwood
  }
`;
exports.schema = schema;
const resolvers = {
  Date: _graphqlIsoDate.GraphQLDate,
  Time: _graphqlIsoDate.GraphQLTime,
  DateTime: _graphqlIsoDate.GraphQLDateTime,
  JSON: _graphqlTypeJson.default,
  JSONObject: _graphqlTypeJson.GraphQLJSONObject,
  Query: {
    redwood: () => ({
      version: _package.default.version,
      currentUser: (_args, context) => {
        return context === null || context === void 0 ? void 0 : context.currentUser;
      }
    })
  }
};
exports.resolvers = resolvers;
